docs/cardano-monads.rst

```diff
@@ -0,0 +1,922 @@
+
+Motivation
+____________________
+
+After a discussion, @georgeee and @int-index came up with a hybrid
+approach that combines many elements of other approaches to provide a
+comprehensive solution.
+
+Design goals (in addition to already discussed criteria):
+
+* avoid data types parametrized by an abstract monad ``m``
+  (including explicit method dictionaries),
+  because it makes reasoning about code complicated
+
+* avoid hard-coded ``IO``, ``MonadIO``, ``MonadBaseControl IO``, etc,
+  in order to have good mocking capabilities and separation of concerns
+
+And here are some observations:
+
+* operationally, our monads can be ``ReaderT ctx IO`` for production code
+  and ``ReaderT ctx (CatchT ST)`` (or similar) in pure tests. Both of these
+  provide (1) passing context, (2) throwing/catching exceptions, (3) mutable
+  state.
+
+* runtime configuration of effects arises for CLI parameters or YAML configuration
+  (for instance, the slotting method is affected by the '--no-ntp' flag). For this
+  use cases effect context sums must be sufficient.
+
+* presumably, any effect dictionary can be separated into its closure (context)
+  and top-level operations
+
+Proposed approach
+_________________
+
+With these facts established, let us consider the approach from multiple angles:
+how effect definitions would look like, and how effect implementatinos look like.
+
+First, let us have a few base monads::
+
+    newtype Prod a = Prod (IO a)
+    newtype TestIO a = TestIO (IO a)
+    newtype TestPure a = TestPure (CatchT ST a)
+
+To define an effect, in the most general case we will need two classes and a type family.
+In the first class we define the methods, mtl-style. The second class is a ``Has``-class to
+extract a part of a context. And a type family determines what context we need for the effect::
+
+    class MonadFoo b where
+        type FooContext b
+        sepulate :: HasFooContext b ctx => ReaderT ctx b Sepulki
+        patanize :: HasFooContext b ctx => ctx
+                                        -> ReaderT ctx b a
+                                        -> ReaderT ctx b (Either Patak a)
+
+    class HasFooContext b ctx where
+        fooContext :: ctx -> FooContext b
+
+A few things to notice:
+
+* the parameter ``b`` stands for "base monad", it will be instantiated to
+  something simple like ``Prod`` or ``TestPure``.
+
+* all methods are defined in the ``ReaderT ctx b`` monad.
+
+* we are fine with this monad being used in a negative position (in
+  ``patanize``), because it is not meant to be lifted/unlifted in any
+  circumstances.
+
+Usage examples
+______________
+
+This is how we could define database-related monads this way::
+
+    class MonadDBGet b where
+        type DBContext b
+        dbGet :: HasDBContext b ctx => DBTag -> ByteString -> ReaderT ctx b (Maybe ByteString)
+
+    class HasDBContext b ctx where
+        dbContext :: ctx -> DBContext b
+        -------- or: Lens' ctx (DBContext b)
+
```

Comment:
`ReaderT cap m` does make a lot of sense. `m` carries the effects, `cap` carries capabilities, i.e. terms implemented over `ReaderT cap m` for that same `cap`. There may be some similarity with free monads: with a free monad you give stand-in symbolic terms for implementations that will be given later; with this approach you directly use implementations that will be given later. It's like skipping a step, inlining the symbolic terms.

I'm just going to write out what I have in mind.

For each capability, there is a record rather than a type class. For example

```Haskell
data Logging cap m = Logging
  { log :: Severity -> Text -> ReaderT (cap m) m ()
  }

data DB cap m = DB
  { getKey :: ByteString -> ReaderT (cap m) m (Maybe ByteString)
  , putKey :: ByteString -> Maybe ByteString -> ReaderT (cap m) m ()
  }
```

The `cap :: (* -> *) -> *` contains implementations of some set of capabilities over a given monad. The capabilities themselves are allowed to use this value, via the reader. This allows for instance a `DB` implementation to use a `Logging` implementation. A constraint would arise which shows how to extract that implementation from the `cap` type:

```Haskell
class Has (capability :: ((* -> *) -> *) -> (* -> *) -> *) (caps :: (* -> *) -> *) (m :: * -> *) where
  take :: caps m -> capability caps m
  replace :: capability caps m -> caps m -> caps m

useCapability
  :: ( Monad m, Has capability caps m )
  => (capability caps m -> ReaderT (caps m) m t) 
  -> ReaderT (caps m) m t
useCapability k = asks take >>= k

logM :: ( Monad m, Has capability caps m ) => Severity -> Text -> ReaderT (caps m) m ()
logM severity text = useCapability (\logging -> log logging severity text)

getKeyM :: ( Monad m, Has capability caps m ) => ByteString -> ReaderT (caps m) m (Maybe ByteString)
getKeyM key = useCapability (\db -> getKey db key)
```

Particular implementations of a capability are just terms of the given record type. Constraints may arise if other capabilities, or effects in the base monad, are required.

```Haskell
-- stdout logging doesn't need any other capabilities (impl is free) but does need
-- MonadIO (m is constrained).
stdoutLogging :: MonadIO m => Logging impl m
stdoutLogging = Logging
  { log = \_ txt -> liftIO (print txt)
  }

-- This stupid DB implementation uses logging, so impl is contrained.
stupidLogDB
  :: ( Monad m, Has Logging impl m )
  => DB (impl m) m
stupidLogDB = DB
  { getKey = \_ -> logM Debug "get" >> return Nothing
  , putKey = \_ _ -> logM Debug "put" >> return ()
  }
```

So ultimately we end up with some `ReaderT (caps m) m t` where there are `Has` constraints on `caps` and `m` and other constraints on `m`. To run the thing, we have to come up with a `caps` type which has all of the necessary capabilities, and a term of that type over an `m` which satisfies all of its constraints.

```Haskell
data MyCapabilities m = MyCapabilities
  { mcLogging :: Logging MyCapabilities m
  , mcDB :: DB MyCapabilities m
  }

instance Has Logging MyCapabilities m where
  take = mcLogging
  replace = ...

instance Has DB MyCapabilities m where
  take = mcDB
  replace = ...
```

A particular implementation of these capabilities doesn't have to specify `m`, it could merely constrain it.

```Haskell
myCapabilities
  :: ( MonadIO m )
  => MyCapabilities m
myCapabilities = MyCapabilities
  { mcLogging = studoutLog
  , mcDB = stupidLogDB
  }
```

Reply:
@georgeee @int-index thoughts?

Reply:
At the first glance I like this direction of travel, could you add it as another proposed approach to the document?

Reply:
Two questions:

1. How can we add capabilities? I would imagine two instances like this:

    ```
    instance {-# OVERLAPPABLE #-} Has cap caps m => Has cap (Ext _ caps) m
    instance Has cap (Ext cap  caps) m
    ```

    then `:*:` could be used for extending the capabilities (where `data Ext f g m = Ext (f m) (g m)`).

2. Do I understand correctly that the `replace` method takes into account the issue of `noLogging` that I outlined in the neighbouring thread?

Reply:
> could you add it as another proposed approach to the document?

Sure.

> Do I understand correctly that the replace method takes into account the issue of noLogging that I outlined in the neighbouring thread?

Yes indeed.

```Haskell
localImplementation
  :: (Has capability caps m)
  => capability caps m
  -> ReaderT (caps m) m t
  -> ReaderT (caps m) m t
localImplementation cap = withReaderT (replace cap)
```

> How can we add capabilities locally? I would imagine two instances like this:

Not necessary. The set of *required* capabilities comes out of the term (GHC infers them). We only have to choose a particular set of capabilities (a `caps m`) when we want to run/interpret the term.

Reply:
> Not necessary. The set of required capabilities comes out of the term (GHC infers them). We only have to choose a particular set of capabilities (a caps m) when we want to run/interpret the term.

Yes, I understand that. I'm talking about the following scenario:

```
f :: (Has cap1 caps m, Has cap2 caps m) => ReaderT (caps m) m r

g :: (Has cap1 caps m) => ReaderT (caps m) m r
g = do
  cap2 <- conjureUpCap2
  withReaderT (Ext cap2) f
```

We want to call `f` from `g`, but `f` needs more capabilities. We construct those capabilities locally and add them to the environment. An example: `SendActions` is a capability that must be constructed on site. By default we have the networking capability, but info about a particular peer becomes available only at a certain point.

Reply:
If the `SendActions` example becomes outdated after the delegation refactoring, the general problem of extending the set of capabilities remains.

Reply:
Oh I see, you want to add *implementations* locally, and have every other implementation which uses that capability rebind to the local implementation. This is basically the same problem as the no-logging case.

So, apparently we need to be able to write terms which abstract over some set of capabilities, such that this set is extensible (the term determines only a subset of it) and such that implementations of these capabilities appear to be dynamically bound (I can set a "local" implementation and all other implementations will use it).

Implicit parameters are unwieldy but nevertheless relevant here. I don't anticipate we'll want to use them but here's how it might look:

```Haskell
data Logging m = Logging
  { log :: String -> m ()
  }

data DB m = DB
  { getKey :: String -> m (Maybe String)
  , putKey :: String -> Maybe String -> m ()
  }

logAndDBExample
  :: ( ?logging :: Logging m
     , ?db :: DB m
     , Monad m
     )
  => m (Maybe String)
logAndDBExample = do
  log ?logging "Writing to DB"
  previous <- getKey ?db "hello"
  putKey ?db "hello" (Just "world")
  return previous

stdoutLogging :: MonadIO m => Logging m
stdoutLogging = Logging
  { log = liftIO . putStrLn
  }

noLogging :: Applicative m => Logging m
noLogging = Logging
  { log = \_ -> pure ()
  }

-- An implementation which uses some other in-scope logging implementation.
prefixLogging :: (?logging :: Logging m) => String -> Logging m
prefixLogging pre = Logging
  { log = \str -> log ?logging (pre ++ str)
  }

-- An implementation which uses the in-scope logging implementation.
stupidLogDB :: ( ?logging :: Logging m, Monad m ) => DB m
stupidLogDB = DB
  { getKey = \key -> log ?logging ("getKey: " ++ key) >> pure Nothing
  , putKey = \key val -> log ?logging ("putKey: " ++ key ++ " -> " ++ show val)
  }

localExample
  :: forall logging db m .
     ( ?logging :: Logging m
     , ?db :: DB m
     , Monad m
     )
  => m (Maybe String)
localExample = do
  logAndDBExample
  -- This does *not* work as we'd like. If we call this with
  -- ?db = stupidLogDB then its ?logger implicit is already bound; we can't
  -- replace it with a new one.
  let ?logging = prefixLogging "prefixed: " in logAndDBExample
  let ?logging = noLogging in logAndDBExample
```

See the comment in `localExample` at the bottom there. Implicit parameters gives extensibility (you just have to choose a canonical name for the capability) but to get @int-index 's requirement that swapping out one implementation should affect all other implementations, I'm not sure it's possible without introducing some other objectionable thing. In the example above, we'd have to move the implicit parameter bindings either into the `getKey` and `putKey` implementations (out of the question because its the implementation (value) not the component (type) which should determine this), or in front of the `DB` term itself (out of the question because, well, look at this):

```Haskell
data DynBind (c :: Constraint) t = DynBind
  { unDynBind :: c => t
  }

dynStupidLogDB :: forall m . Monad m => DynBind (?logging :: Logging m) (DB m)
dynStupidLogDB = DynBind stupidLogDB

dynLogAndDBExample
  :: forall logging dyndb .
     ( ?logging :: Logging IO
      -- Try to make it work without writing the ?logging binding here in the DynBind constraint.
      -- I think it's impossible.
      -- But the fact that we have to write it is huge points against this solution. This term should
      -- work regardless of the bindings that dyndb uses.
     , ?dyndb :: DynBind (?logging :: Logging IO) (DB IO)
     )
  => IO (Maybe String)
dynLogAndDBExample =
  let ?db = unDynBind ?dyndb in do
    log ?logging "Writing to DB"
    previous <- getKey ?db "hello"
    let ?logging = noLogging :: Logging IO in let ?db = unDynBind ?dyndb in putKey ?db "hello" (Just "world")
    return previous
```

Reply:
Let's think about the semantics of these local implementations. I think `SendActions m` is a non-example but whatever, it's been brought up, so let's use it.

If I give a local implementation of a `SendActions m`, it can use the existing capabilities, but of course the existing capabilities can't use it (if they were aware of this capability, then it would appear as a constraint, and we'd just be overriding it). This seems to imply that it's impossible to give mutually independent local implementations. With a solution in the style of `Ext` in @int-index 's comment above, neither of the mutually independent things can be added first, they must be added to the environment together. So we have to add a composite thing with `Has` instances of its own and we end up with duplicate instances

```Haskell
instance Has capability local => Has capability (Extend local global)
instance Has capability global => Has capability (Extend local global)
```

This gets really complicated, really fast, so perhaps the most important question in this discussion is "are we absolutely sure that this is necessary?" Would it be so inconvenient to drop it? Are there compelling real-world use cases for swapping in a local implementation with the aforementioned semantics? Is it no sufficient to have one application-global set of abstracted capabilities that is fulfilled once and only once, i.e. when the application in run?

Reply:
> Are there compelling real-world use cases for swapping in a local implementation with the aforementioned semantics? Is it not sufficient to have one application-global set of abstracted capabilities that is fulfilled once and only once, i.e. when the application in run?


We do have use cases for extending the set of capabilities locally — see `runToilAction`. Right now we do it by adding a transformer on top of the base monad, but we'd like to avoid transformers in the new approach. We use `DBToil` to add `MonadUtxoRead` and `MonadStakesRead` capabilities, and we use `ToilT` (which is a `StateT`, unfortunately — must be changed!) to modify the behavior of `MonadUtxoRead`, `MonadStakesRead`, `MonadUtxo`, and `MonadStakes`.

Furthermore, after your comments I've just realized that if we allow extending, we automatically allow problematic overrides. And it might be the case this sort of bugs (which `replace` avoids) is already lurking somewhere in our toil-related code.

It'd be really nice to come up with a solution that would let us stop worrying about all this.

An idea that comes to my mind is to pass all capabilities in a map that keeps track of available capabilities at type level:

```
newtype Capabilities (caps :: '[(* -> *) -> *]) (m :: * -> *) = Capabilities (Map TypeRep Any)

We could have `caps :: Capabilities '[Logging, Networking] m` with two elements: `(typeRep (Proxy :: Proxy Logging), logging)` and `(typeRep (Proxy :: Proxy Networking), networking`.

Adding/overriding capabilities becomes a single operation:

```
toAny :: a -> Any
toAny = unafeCoerce -- as usual

insertCap :: forall cap m. cap m -> Capabilities caps m -> Capabilities (cap ': caps) m
insertCap cap (Capabilities m) = Capabilities $ Map.insert (typeRep (Proxy :: Proxy cap)) (toAny cap) m
```

And extracting a capability can be done like this:

```
type family Elem x xs :: Constraint where
  Elem x (x ': xs) = ()
  Elem x (y ': xs) = Elem x xs
  Elem x '[] = TypeError ... -- user-defined type errors are neat

fromAny :: Any -> a
fromAny = unsafeCoerce -- as usual

lookupCap :: forall cap m caps. Elem cap caps => Capabilities caps m -> cap m
lookupCap (Capabilities m) = fromAny . fromJust . Map.lookup (typeRep (Proxy :: Proxy cap)) m
```

The nice thing about this solution is that we can arbitrarily extend the `Capabilities` map, and still be sure that each capability is represented by a single value inside — that's how `Map` works operationally. Needless to say that `Capabilities` must be an abstract type in a separate module, and only `insertCap` and `lookupCap` exported for it.

This *might* look like too much trickery, but we only have to do it once.

Reply:
> We do have use cases for extending the set of capabilities locally — see runToilAction. Right now we do it by adding a transformer on top of the base monad, but we'd like to avoid transformers in the new approach. We use DBToil to add MonadUtxoRead and MonadStakesRead capabilities, and we use ToilT (which is a StateT, unfortunately — must be changed!) to modify the behavior of MonadUtxoRead, MonadStakesRead, MonadUtxo, and MonadStakes.

I've only given a quick glace over the transaction toil pieces. If I understand correctly, this concerns verifying and applying transactions and blocks, and depends upon some state (the utxo for one). I noticed a warning in a comment: "this function may apply some transactions and fail eventually. Use it only on temporary data".

I see `MonadUtxo m`, `MonadGState m`, `MonadStake m`, `MonadTxPool m`, `WithLogger m` in the various toil modes. Does `ToilT` really *modify* the behaviour of these things? Doesn't it simply make use of them?

I wonder, could this not be made pure? Read in all the required state up front (feasible?) and then run a pure (possibly `StateT`, `ExceptT`, whatever, as long as its over `Identity` base) computation against that state. A caller can then update the relevant state if it succeeds.

Reply:
> Furthermore, after your comments I've just realized that if we allow extending, we automatically allow problematic overrides. And it might be the case this sort of bugs (which replace avoids) is already lurking somewhere in our toil-related code.
>
> It'd be really nice to come up with a solution that would let us stop worrying about all this.

Yes! :) This really ought to be a controversial feature. Google "late binding" and read about all of the love *and* hate. What we're discussing here feels a lot like late binding although technically it's not. 

> The nice thing about this solution is that we can arbitrarily extend the Capabilities map, and still be sure that each capability is represented by a single value inside — that's how Map works operationally. Needless to say that Capabilities must be an abstract type in a separate module, and only insertCap and lookupCap exported for it.
>
> This might look like too much trickery, but we only have to do it once, perhaps in a separate lib, and the usage is straightforward.

To get this late/dynamic binding style, each of the capabilities in the map would have to know the set of capabilities that will be "dynamically" provided (i.e. come from the reader context). That's why we see the `cap` type in the example record types:

```Haskell
-- cap is the set of all capabilities available, so that a `log` implementation is able to use
-- other (dynamically bound) capabilities.
data Logging cap m = Logging
  { log :: Severity -> Text -> ReaderT (cap m) m ()
  }
```

Fitting this into the map scheme will be awkward I think. But I should note that if we give up type safety (!!) then everything is easy :) Just use a `Map String Any` in the reader context and hope for the best. But I'm sure we can agree not to do that.

Reply:
> Fitting this into the map scheme will be awkward I think

Not really, I think we can just make the `Map` defined like this:

```
newtype Capabilities (caps :: '[(* -> *) -> *]) (m :: * -> *) =
  Capabilities (Map TypeRep (Capabilities caps m -> Any))
```

so instead of `Any` representing the dict, it's a function from all of the capabilities to the dict. The fact that I did not include this is just an oversight, I really intended this to be a fixed point.

> Just use a Map String Any in the reader context and hope for the best. But I'm sure we can agree not to do that.

Just to clarify: I intend the unsafe interface to be hidden behind a type-safe one. End-users won't deal with `Any`.

> I wonder, could this not be made pure? 

I'm not familiar with this code enough to say. If it could be made pure, of course that'd be preferable, but it's not the only use case. In LRC currently we do something quite awkward — we apply blocks, see the result, and rollback, using the real database (because copying it into memory would be to inefficient). This requires us to bypass the `k` security parameter, and also it's dangerous if the application is killed during LRC and the base is left in non-rollbacked state. Instead we could try to use a DB adapter locally, which would rely on the real DB for the bulk of data, but also keep track of temporary changes.

I think we might find more use cases if we dig further. Point is: "late binding" seems useful, the lack of it seems bug-prone.

I am going to try to put together a package with the Capabilities map, and see if it works out. Then we'll have concrete code to discuss.

Reply:
> I'm not familiar with this code enough to say. If it could be made pure, of course that'd be preferable, but it's not the only use case. In LRC currently we do something quite awkward — we apply blocks, see the result, and rollback, using the real database (because copying it into memory would be to inefficient). This requires us to bypass the k security parameter, and also it's dangerous if the application is killed during LRC and the base is left in non-rollbacked state. Instead we could try to use a DB adapter locally, which would rely on the real DB for the bulk of data, but also keep track of temporary changes.

I'm horrified, and I'm sure there is some better solution. Is there too much data to copy into memory at once? Morally the LRC computation should be pure, and the impure pieces surrounding it would not need to know of any "LRC capability". Based on this summary you've given, I don't think this is a compelling use case. It does not strike me as something we should aim to support, it's something we should aim to correct.

> I think we might find more use cases if we dig further. Point is: "late binding" seems useful, the lack of it seems bug-prone.

Perhaps you should expand on this. I don't appreciate your point. Do the existing ToilT and LRC actually depend upon a dynamic/late binding style local override for effects? Even if they do, from what I've read, they are poorly implemented anyway, so any argument staked on them is a rather weak one.

Reply:
>  Is there too much data to copy into memory at once?

Yes, I think @gromakovsky or @pva701 can provide more details, I've only touched this code a little (and was as horrified as you).

> Do the existing ToilT and LRC actually depend upon a dynamic/late binding style local override for effects? 

`ToilT` was written without early/late binding in mind, it was written without thinking about this issue at all. It could be buggy because of this, but I think the more natural inclination is to assume late binding (I hoped that logging/networking example is illustrative, and I had to deal with a similar problem in `Auxx` when I tried to change the logging method there). This is why I'm saying the lack of late binding seems more bug prone: I bet when a programmer tries to override a capability locally, he intends to do it fully (you could argue that this depends on the programmer, but it's true at least for me).

As to LRC, it does not rely on late binding currently. I'm saying that in order to rewrite the current horrible implementation, we could make a good use of late binding.

Reply:
> This is why I'm saying the lack of late binding seems more bug prone: I bet when a programmer tries to override a capability locally, he intends to do it fully (you could argue that this depends on the programmer, but it's true at least for me).

On the other hand, if I have a db capability, I don't expect changing my logging capability to alter the behaviour of the db capability. The implementation I gave above expresses this in a pure functional way, but it doesn't *appear* to be pure. The dependence on the global vtable in the `ReaderT` makes for a programming style that seems impure: there's no clear and apparent dependence of the db on the logging capability, yet changing the latter changes the former. It feels like updating a global variable.

> ToilT was written without early/late binding in mind, it was written without thinking about this issue at all. It could be buggy because of this

I highly doubt that's the reason. My hunch is that it's not pure enough, i.e. it's too much exposed to the issues we're discussing.

I'm trying to imagine a case where something doesn't work right only because the wrong implementation for some effect was chosen. Is it the fault of the definition? No, it's correct assuming the effects are implemented correctly. So it's the fault of some caller, who was apparently confused about how to specify which implementation shall be used for an effect. What's the simplest most straightforward way to choose effects? Function application! The best tool for abstraction we have is and probably always will be lambda. What's more confusing: a higher-order function, or a type-indexed `Any`-valued `Constraint` map keyed on `TypeRep`s that's antecedent to every effect implementation?

***

I think the desire for dynamic binding could be the symptom of a deeper issue: there's not enough purity in cardano-sl.

There's a trend in cardano-sl to mix complex pure terms with complex effectful terms. The appearance of an `IO` in the type is alarming to any Haskell programmer, so the adopted solution was apparently to relax it to an `m`, and constrain it with classes to implement all of the impure things, and so we are where we are now.

But there's another option: maximally factor out all of the pure parts, and use them to implement small effectful parts. The pure parts are by nature agnostic to your choice of database or logging or whatever, but if they're mixed in with the effectful parts, it *appears* as though they are *not*, and so the *perceived* benefit of these constraints on `m` is inflated: look how much code we're re-using! But all of the pure parts are reusable across effect implementations even without this extra machinery, so what it's really bought you is far far less.

Now suppose you want to use a different logging implementation, and you want your db implementation to also use that new logging implementation. Instead of being confronted with some gargantuan

```Haskell
processTheTransactions :: ( SuperWorkMode m ) => m AllDoneUpdatedSomeStateForYou
```

you might find the more versatile

```Haskell
pureProcessTransactions :: Set Tx -> Either Invalid TxPayload
```

and you can just write out what you want because it's so simple

```Haskell
let logging = myExplicitLoggingImpl
    db = myExplicitDbImpl logging
in  do txs <- getTxs db
       log logging "Got txs"
       pure $ pureProcessTransactions txs
```